python/python_direct/python_translate.cpp

```diff
@@ -0,0 +1,506 @@
+      }
+    }
+    return false;
+  }
+
+  // Check that all of the expression's inputs are defined in FusionDefinition.
+  bool checkExpressionDependencies(Expr* e) {
+    // TODO Add check_view_dependency
+    return std::all_of(
+        e->inputs().begin(), e->inputs().end(), [&](const Val* v) {
+          return visited_vals_.count(v) > 0;
+        });
+  }
+
+  void translate() {
+    printer_.generateFusionDefinition();
+
+    // Add Fusion inputs to FusionDefinition
+    for (nvfuser::Val* v : fusion_->inputs()) {
+      dispatch(v);
+    }
+
+    // Gather all expressions in CPP Fusion.
+    const std::vector<nvfuser::Expr*> fusion_exprs = fusion_->exprs();
+    std::deque<nvfuser::Expr*> to_visit(
+        fusion_exprs.begin(), fusion_exprs.end());
+
+    // TODO: Scalar expressions are not handled by Fusion::exprs, so gather them
+    // manually.
+
+    // Topological search of Fusion expressions
+    size_t skip_count = 0;
+    std::unordered_set<nvfuser::Expr*> visited;
+    while (!to_visit.empty()) {
+      Expr* e = to_visit.front();
+      to_visit.pop_front();
+
+      NVF_ERROR(
+          skip_count <= to_visit.size(),
+          "Cycle detected: None of the expressions can be processed!");
+
+      // short-circuit: skip if already visited
+      if (visited.count(e) > 0) {
+        continue;
+      }
+
+      // TODO: short-circuit: skip Split and Merge expressions created by
+      // Reshape
+      // TODO: short-circuit: skip Resize expressions created by Slice
+
+      bool is_expr_inputs_valid =
+          std::all_of(e->inputs().begin(), e->inputs().end(), [this](Val* v) {
+            return !v->isA<TensorView>() ||
+                !isScheduledTensorView(v->as<TensorView>());
+          });
+      NVF_ERROR(
+          is_expr_inputs_valid,
+          "Found a TensorView with scheduled loop domain.");
+
+      // Handle scalars and constants not generated by separate expression.
+      std::vector<Val*> scalars;
+      std::copy_if(
+          e->inputs().begin(),
+          e->inputs().end(),
+          std::back_inserter(scalars),
+          [](Val* v) { return v->isScalar(); });
+      std::for_each(scalars.begin(), scalars.end(), [this](const Val* v) {
+        dispatch(v);
+      });
+
+      // short-circuit: add to back of stack if not all of the expression's
+      // dependencies are satisfied.
+      if (!checkExpressionDependencies(e)) {
+        ++skip_count;
+        to_visit.push_back(e);
+        continue;
+      }
+
+      // Create RecordFunctor given inputs, outputs, and attributes.
+      visited.insert(e);
+      dispatch(e);
+      skip_count = 0;
+    }
+
+    // Add tensor outputs and handle aliased outputs
+    std::unordered_set<nvfuser::Val*> visited_alias_output;
+    for (nvfuser::Val* v : fusion_->outputs()) {
+      NVF_ERROR(v->isA<TensorView>());
+      const AliasInfo& alias_info = fusion_->getOutputAlias(v);
+      switch (alias_info.type) {
+        case AllocationType::New: {
```

Comment:
Do we need any special treatment of  AllocationType::Evaluate? I thought it's just the same as AllocationType::New